File: src/FunCrossMark.f
C
C     Author: Carles Comas
C     Date: 3th FEBRUARY 2023 
C
C     Function  Functional.mark.cor.f to obtain the Functional Mark Correlation Function (FMCF)
C     
C

       SUBROUTINE funcrossmarksub(f1,f2,d,n,r,nr,nrf,minrf,maxrf,
     &                 kr,corr,bw,h,lamd,cor,edge,Ar,FMCF)


       IMPLICIT real*8 (a-h,o-z)

C      PARAMETERS
C      i, j, i1 internal counting variables, integer
C      TestF, matrix with Test function combining functions of the point pattern, test function: L2 distances
C      d, pairwise distance. double precision d(n,n)
C      n, number of points of the point pattern, interger
C      r, sequence of values, for the discret set of distances where the FMCF is evaluated, double precision
C      nr, the length of r, integer
C      nrf, the legnth of rf, integer
C      kr, value of the kernel function, integer value
C      bw, bandwidth value, real*8
C      lamd, point intensity, double precision
C      cor, value of the isotropic correction for point i wrt point j
C      edge, value for the edge-effect correcion, interger 
C      Ar, area of the window of observation, double precision
C      FMCF, Functional Mark Correlation Function, output

       INTEGER  i, j, i1, n, nr, kr, corr, edge, nrf, nn  
       DOUBLE PRECISION d(n,n), r(nr), lamd(n), cor(n,n)
       DOUBLE PRECISION TestF(n,n), FMCF(nr), PCF(nr)
       DOUBLE PRECISION fxs(nrf), f1(nrf,n)
       DOUBLE PRECISION maxrf(n,n), f2(nrf,n), minrf(n,n)
       REAL*8 Ar, bw, dij, kerns, ExpF, a, b, h
       DIMENSION kr(3), edge(2),corr(2)
     
       
       
     


C      PAIR CORRELATION FUNCTION

       DO i1=1, nr
        PCF(i1)=0.0
         DO i=1,n
          DO j=1,n
           IF (j.ne.i) THEN
            dij=d(i,j)
              IF (kr(1).eq.1) THEN
c             kerns=1
               kerns=boxkernel((r(i1)-dij)/bw,bw)
              ELSE IF (kr(2).eq.1) THEN
               kerns=ekernel((r(i1)-dij)/bw,bw)
              ELSE IF (kr(3).eq.1) THEN
              kerns=qkernel((r(i1)-dij)/bw,bw)
              END IF

              IF (kerns.ne.0) THEN
C     none   
                IF (edge(1).eq.1) THEN
                 wij=kerns/(lamd(i)*lamd(j)*r(i1))
                 PCF(i1)=PCF(i1)+wij 
                END IF                          
C    isotropic
                IF (edge(2).eq.1) THEN                  
                 wij=(kerns*cor(i,j))/(lamd(i)*lamd(j)*r(i1))
                 PCF(i1)=PCF(i1)+wij
                END IF

              END IF
           END IF
         END DO
        END DO
        PCF(i1)=PCF(i1)/(2*3.141592654*Ar)
       END DO


       DO i=1, n
         DO j=1, n
          TestF(i,j)=0.0
         ENDDO
       ENDDO

C   TEST FUNCTION PAPER TEST 2009

C        DO i=1, n
C         DO j=1, n
C           IF (j.ne.i) THEN 
C             DO i1=1, nrf
C             fxs(i1)=(f1(i1,i)-f1(i1,j))**2
C             ENDDO
C             TestF(i,j)=sqrt(SIMP(fxs, minrf(i,j), maxrf(i,j),nrf))
C           ENDIF
C         ENDDO
C        ENDDO


c    TEST FUNCTIONS

C   FUNCTIONAL MARK VARIOGRAM FUNCTION

      IF (corr(1).eq.1) THEN
       DO i=1, n
           DO j=1, n
           IF (j.ne.i) THEN 
            DO i1=1, nrf
              fxs(i1)=0.5*(f1(i1,i)-f2(i1,j))**2
            ENDDO
            TestF(i,j)=SIMP(fxs, minrf(i,j), maxrf(i,j),nrf,h)
           ENDIF
           ENDDO
          ENDDO
      ENDIF

C   FUNCTIONAL MARK CORRELATION FUNCTION
      IF (corr(2).eq.1) THEN
        DO i=1, n
           DO j=1, n 
           IF (j.ne.i) THEN 
            DO i1=1, nrf
              fxs(i1)=f1(i1,i)*f2(i1,j)
            ENDDO
             TestF(i,j)=SIMP(fxs, minrf(i,j), maxrf(i,j),nrf,h) 
            ENDIF
           ENDDO
         ENDDO
      ENDIF

   

C   MEAN FUNCTIONAL MARK CORRELATION FUNCTION
C      IF(corr(2).eq.1) THEN
      ExpF=ETestF(n,TestF)
C      ENDIF

C      IF(corr(1).eq.1) THEN
C      ExpF=1.0
C      ENDIF

C   GENERAL FUNCTIONAL MARK CORRELATION FUNCTIONS

      DO i1=1, nr
        FMCF(i1)=0.0
         DO i=1,n
          DO j=1,n
           IF (j.ne.i) THEN
            dij=d(i,j)
              IF (kr(1).eq.1) THEN
               kerns=boxkernel((r(i1)-dij)/bw,bw)
              ELSE IF (kr(2).eq.1) THEN
               kerns=ekernel((r(i1)-dij)/bw,bw)
              ELSE IF (kr(3).eq.1) THEN
              kerns=qkernel((r(i1)-dij)/bw,bw)
              END IF

              IF (kerns.ne.0) THEN
C     none   
                IF (edge(1).eq.1) THEN
             wij1=(kerns*TestF(i,j))
             wij2=lamd(i)*lamd(j)*r(i1)*ExpF*PCF(i1)
             wij=wij1/wij2     
                 FMCF(i1)=FMCF(i1)+wij 
                END IF                          
C    isotropic
                IF (edge(2).eq.1) THEN                  
            wij1=kerns*TestF(i,j)*cor(i,j)
            wij2=lamd(i)*lamd(j)*r(i1)*ExpF*PCF(i1)
            wij=wij1/wij2
                 FMCF(i1)=FMCF(i1)+wij
                END IF

              END IF
           END IF
         END DO
        END DO
        FMCF(i1)=FMCF(i1)/(2*3.141592654*Ar)
       END DO


        RETURN
        END
       

       
CCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCc
C
C     functions called by :
C     -----------------------------------------
C
C     * boxkernel, ekernel, qkernel
C
CCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCC

C--------------------------------------------------------------------
C
C     boxkernel
C
C--------------------------------------------------------------------

       FUNCTION boxkernel(x,h)

       IMPLICIT REAL*8 (a-h,o-z)

       REAL*8 x, h

       IF (abs(x).le.1d0) THEN
         boxkernel=1d0/2d0
       ELSE
         boxkernel=0d0
       END IF
       boxkernel=boxkernel/h
       RETURN
       END

C--------------------------------------------------------------------
C
C     Epanechnikov kernel
C
C--------------------------------------------------------------------

       FUNCTION ekernel(x,h)

     
       IMPLICIT REAL*8 (a-h,o-z)
       REAL*8 x, h

       IF (abs(x).le.1) THEN
           ekernel=(3d0/(4d0*h))*(1-x**2)
       ELSE
          ekernel=0.0
       ENDIF

C       IF (abs(x).le.1d0) THEN
C           ekernel=(3d0/4d0)*(1-x**2)
C       ELSE
C           ekernel=0d0
C       END IF
C       ekernel=ekernel/h

        RETURN
        END

 

C--------------------------------------------------------------------
C
C     quartic (biweight) kernel
C
C--------------------------------------------------------------------

       FUNCTION qkernel(x,h)

       IMPLICIT REAL*8 (a-h,o-z)
       REAL*8 x, h

       IF (abs(x).le.1d0) THEN
           qkernel=(15d0/16d0)*(1-x**2)**2
       ELSE
           qkernel=0d0
       END IF
       qkernel=qkernel/h

       RETURN
       END

C--------------------------------------------------------------------
C
C     Expected Test Function
C
C--------------------------------------------------------------------

      FUNCTION ETestF(n,TestF)

      IMPLICIT REAL*8 (a-h,o-z)
      INTEGER n
      DOUBLE PRECISION TestF(n,n)

       ETestF=0.0
       DO i=1,n
          DO j=1,n
c           IF (j.ne.i) THEN
           ETestF=ETestF+TestF(i,j)
c           ENDIF
         ENDDO
       ENDDO
       ETestF=ETestF/(n**2)
       RETURN
       END


C--------------------------------------------------------------------
C
C     Integral Function using Simpson 1/3 rule
C
C--------------------------------------------------------------------
        FUNCTION SIMP(fxs,a,b,nrf,h)
        IMPLICIT REAL*8 (a-h,o-z)
        INTEGER nrf, i2, i3
        DOUBLE PRECISION fxs(nrf),coef(nrf)
        REAL*8 h, ni, sum1, a, b

       
        DO i=1,nrf
           coef(i)=1
        ENDDO    
  
         SIMP=0.0
        IF(a.lt.(b-3)) THEN
        ni=b-a+1
c        h=(b-a)/(ni-1)
       
        sum1=0


        IF (MOD(ni,2.0) == 0) THEN

          i2=a+1
          i3=b-2
           DO i=i2,i3
             IF (MOD(i,2)== 0) THEN
                coef(i)=4
             ELSE
                coef(i)=2
             ENDIF
            sum1=coef(i)*fxs(i)+sum1
           ENDDO 
           i2=b-1
           i3=b
           i4=a
            SIMP=h/3*(sum1+fxs(i2)+fxs(i4))+h/2*(fxs(i2)+fxs(i3))
         ELSE
      
          i2=a+1
          i3=b-1
          i4=a
          i5=b
          DO i=i2,i3
             IF (MOD(i,2)== 0) THEN
                coef(i)=4
             ELSE
                coef(i)=2
             ENDIF
            sum1=coef(i)*fxs(i)+sum1
          ENDDO 
           SIMP=h/3*(sum1+fxs(i5)+fxs(i4))
         ENDIF
        ENDIF

         SIMP=SIMP*1

c          IF(SIMP == SIMP) THEN
c             SIMP=SIMP*1
c          ELSE
c             SIMP=0.0
c          ENDIF

         RETURN
         END
       


c        FUNCTION SIMP(fxs,a,b,nrf)
c        IMPLICIT REAL*8 (a-h,o-z)
c        INTEGER nrf,c1,c2
c        DOUBLE PRECISION fxs(nrf)
c        REAL*8 a, b, h
        
c         h=(b-a)/(nrf-1.0)	
c         SIMP=3.0*(fxs(1)+fxs(nrf))/8.0+7.0*(fxs(2)+
c     &    fxs(nrf-1))/6.0+23.0*(fxs(3)+fxs(nrf-2))/24.0
c         c1=4
c         c2=nrf-3
c         DO i=c1, c2
c          SIMP=SIMP+fxs(i)
c         ENDDO
c         SIMP=SIMP*h
c         RETURN
c         END

File: R/FunCrossMark.R
#' Function to obtain the Auto- and cross-function mark summary characteristics for the mark variogram and mark correlation
#'
#' This function computes an estimator of this function as described in Eckardt, Comas and Mateu (2023)
#'
#' @examples 
#'  using Data_Duke
#'  Functiona.mark.cor(
#'  
#' @param xy, object of class ppp (point pattern)
#' @param r, sequence of values to evaluate the Auto- and cross-function mark summary characteristics
#' @param rf, sequence of values for the function associated to each point
#' @param f1 is a matrix that contains the first set of functions for each point, funct1[,1] is the function of set 1 for point 1
#' @param f2 is a matrix that contains the second set of functions for each point, funct1[,1] is the function of set 2 for point 1
#' @param kernel, kernel function for both the pair correlation function and the function mark summary characteristic
#' @param TestFunc, Test function. This can be TestFunc="variogram" or TestFunc="correlation"
#' @param bw bandwidth for the kernel function for the pair correlation function and Functional mark correlation function
#' @param lambda, point intensity
#' @param stoyan value for the epanechnikov kernel, default stoyan=0.15
#' @param correction for edge-effect for both the pair correlation function and the Functional mark correlation function
#' @return 
#' @author Carles Comas \email{carles.comas@udl.cat}
#' @useDynLib SppFDA, .registration=TRUE
#' @import spatstat
#' @export

FunCrossMark <- function(xy, r, rf, f1, f2, stoyan=0.15, TestFunc = "variogram", kernel = "epanechnikov", bw, lambda, correction = "isotropic"){
       
         verifyclass(xy, "ppp")
         correc <- c("none", "isotropic")
        
         if (missing(f1)) stop("the f1 matrix should be given")
         if (!inherits(f1, "matrix")) stop("funct1 should be a matrix")
         if (missing(f2)) f2<-f1
     
        

         if (missing(rf)) stop("the rf sequence should be given")
         if (!inherits(rf, "numeric")) stop("rf should be a numeric valuess")
         if (length(rf)<7) stop("rf should be a sequence of values larger than 6")

         indc <- match(correction, correc, nomatch = NA)
         if (any(cm <- is.na(indc))){
          nocm <- paste("unrecognised correction method:",paste(dQuote(correction[cm]),collapse=","))
          stop(nocm,call. = FALSE)
         }
         
         indc <- unique(indc)	
         edge <- rep(0, 2)
         edge[indc] <- 1	
         
         corr <- c("variogram", "correlation")
         indr <- match(TestFunc, corr, nomatch = NA)
         if (any(kmr <- is.na(indr))){
         nokmr <- paste("unrecognised test function:",paste(dQuote(TestFunc[kmr]),collapse=","))
         stop(nokmr,call. = FALSE)
         }
         indr <- unique(indr)
         corr2 <- rep(0, 2)
         corr2[indr] <- 1

         kern <- c("rectangular", "epanechnikov", "biweight")
         indk <- match(kernel, kern, nomatch = NA)
         if (any(km <- is.na(indk))){
         nokm <- paste("unrecognised kernel function:",paste(dQuote(kernel[km]),collapse=","))
         stop(nokm,call. = FALSE)
         }

         indk <- unique(indk)
         ker2 <- rep(0, 3)
         ker2[indk] <- 1

         win <- xy$window 
         n <- xy$n
         areaW <- area.owin(win)

         if(missing(lambda)) lambda <- n/areaW

         if (missing(bw) && (kernel == "epanechnikov") && length(lambda)== 1) {
          bw <- stoyan/sqrt(lambda)
         }
 
         if (missing(bw)){
            if((kernel != "epanechnikov") | length(lambda)>1) {
           bw <- bw.pcf(X = xy, kernel = kernel)[1] 
            }
         }
         if (!inherits(bw, "numeric")) stop("bw should be a numeric value")

        

         if (missing(r)){
         rect <- as.rectangle(win)
         maxd <- min(diff(rect$xrange), diff(rect$yrange))/4
         r <- seq(maxd*0.1, maxd, len = 301)[-1]
         r <- sort(r)
         }

         if(r[1] == 0) r <- r[-1]
         if (!inherits(r, "numeric")) stop("r should be a numeric value")
         if (length(r)<=1) stop("r should be a sequence of values")

         kern <- c(kernel = kernel, bw = bw)
         
         n <- xy$n
         il <- seq(1, n)
         nr <- length(r)
         areaW <- area.owin(win)
         Ar<-areaW

       
         if (length(lambda) == 1) lambda <- rep(lambda, n)
    
         d <- pairdist(xy)
         if(correction == "isotropic") cor <- edge.Ripley(xy, d)
         if(correction == "none") cor <- 1

       
          nrf<-length(rf)

         na11<-c()
          for(i in 1:n){ 
            a11<-f1[,i]
            na11[i]=(nrf+1)-length(a11[!is.infinite(a11)])
          }

          na12<-c()
          for(i in 1:n){ 
            a12<-f2[,i]
            na12[i]=(nrf+1)-length(a12[!is.infinite(a12)])
          }


          minrf<-array(NA,c(n,n))
          dim(minrf)=c(n,n)

          for(i in 1:n){
           for(j in 1:n){
            na<-max(na11[i],na12[j])
            minrf[i,j]<-na
           }
          }

       
            na21<-c()
          for(i in 1:n){ 
            a21<-f1[,i]
            na21[i]=length(a21[!is.na(a21)])
          }

          na22<-c()
          for(i in 1:n){ 
            a22<-f2[,i]
            na22[i]=length(a22[!is.na(a22)])
          }

          
          maxrf<-array(NA,c(n,n))
          dim(maxrf)=c(n,n)

          for(i in 1:n){
           for(j in 1:n){
            na3<-min(na21[i],na22[j])
             maxrf[i,j]<-na3
           }
          }  


        f1[is.na(f1)] <- 1.0
        f1[is.infinite(f1)] <- 1.0
        f2[is.na(f2)] <- 1.0
        f2[is.infinite(f2)] <- 1.0
         h<-rf[2]-rf[1]
       
         FMCF<-rep(10,nr)
         storage.mode(FMCF) <- "double"
                   
out<-.Fortran("funcrossmarksub", f1=as.double(f1), f2 = as.double(f2),
                       d = as.double(d), n = as.integer(n), r = as.double(r),
                       nr = as.integer(nr), nrf=as.integer(nrf), minrf=as.double(minrf),
                       maxrf=as.double(maxrf), ker2 = as.integer(ker2),corr2 = as.integer(corr2),
                       bw = as.double(bw), h = as.double(h),lambda = as.double(lambda),
                       cor = as.double(cor), edge = as.integer(edge),
                       Ar=as.double(Ar), out=(FMCF), PACKAGE = "SppFDA")
                      
invisible(return(list(FMCF = out$out, r = r, kernel = kern, lambda = lambda)))
}











